src/Instrumentation/CakePHP/src/Hooks/Cake/Http/Server.php

```diff
@@ -0,0 +1,199 @@
+                    $this->metricAttributes += [TraceAttributes::ERROR_TYPE => get_class($exception)];
+                }
+                if ($response) {
+                    if ($response->getStatusCode() >= 400) {
+                        $span->setStatus(StatusCode::STATUS_ERROR);
+                        $this->metricAttributes += [TraceAttributes::ERROR_TYPE => $response->getStatusCode()];
+                    }
+
+                    $span->setAttribute(TraceAttributes::HTTP_RESPONSE_STATUS_CODE, $response->getStatusCode());
+                    $span->setAttribute(TraceAttributes::NETWORK_PROTOCOL_VERSION, $response->getProtocolVersion());
+                    $span->setAttribute(TraceAttributes::HTTP_RESPONSE_BODY_SIZE, $response->getHeaderLine('Content-Length'));
+                    $this->metricAttributes += [TraceAttributes::HTTP_RESPONSE_STATUS_CODE => $response->getStatusCode()];
+                }
+
+                $span->end();
+
+                if($this->isRoot()) {
+                    $this->recordDuration($span, $this->metricAttributes);
+                    $this->recordRequestBodySize($request, $this->metricAttributes);
+                    $this->recordResponseBodySize($response, $this->metricAttributes);
+                }
+            },
+        );
+    }
+
+    private function extractHostAddressAndPort(ServerRequestInterface $request): array
+    {
+        $scheme = $request->getUri()->getScheme();
+        $defaultPort = $scheme === 'https' ? 443 : 80;
+
+        $xForwardedHost = $request->getHeader('X-Forwarded-Host');
+        $xForwardedHost = empty($xForwardedHost) ? null : $xForwardedHost[0];
+        if($xForwardedHost !== null) {
+            return $this->parseHostString($xForwardedHost, $defaultPort);
+        }
+
+        $forwarded = $request->getHeader('Forwarded');
+        $forwarded  = empty($forwarded) ? null : $forwarded[0];
+        if($forwarded !== null) {
+            $parts = explode(';', $forwarded);
+            $host = preg_grep('/^host=.*$/', $parts);
+            if ($host) {
+                return $this->parseHostString($host[0], $defaultPort);
+            }
+        }
+
+        $hostHeader = $request->getHeader('Host');
+        $hostHeader = empty($hostHeader) ? null : $hostHeader[0];
+        if ($hostHeader) {
+            return $this->parseHostString($hostHeader, $defaultPort);
+        }
+
+        return [$request->getUri()->getHost(), $defaultPort];
+    }
+
+    private function parseHostString(string $host, int $defaultPort): array
+    {
+        $parts = explode(':', $host);
+        if(count($parts) < 2) {
+            return [$parts[0], $defaultPort];
+        }
+
+        return [$parts[0], $parts[1]];
+    }
+
+    /**
+     * @param SpanInterface $span
+     * @param array $metricAttributes
+     * @return void
+     */
+    private function recordDuration(SpanInterface $span, array $metricAttributes): void
```

Comment:
this type of thing is usually left to post-processing by something like https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/connector/spanmetricsconnector/README.md
Also, for a shared-nothing setup (apache, fcgi, etc) I think you're likely to run into https://github.com/open-telemetry/opentelemetry-php/issues/970#issuecomment-1513427738

Reply:
Thank you for pointing this out. I wasn't aware of the problems with shared-nothing setups. 

Just to make sure I understand this correctly: You would also create the metrics for request and response body size in post-processing or using something like the upcoming [TraceQL metrics](https://grafana.com/docs/tempo/latest/traceql/metrics-queries/) of Grafana Tempo?

In that case, I will remove all the metrics related code.

Reply:
> Just to make sure I understand this correctly: You would also create the metrics for request and response body size in post-processing or using something like the upcoming [TraceQL metrics queries](https://grafana.com/docs/tempo/latest/traceql/metrics-queries/) of Grafana Tempo?

Yes. If your root span contains those attributes, then you can add more dimensions to the span metrics connector to capture them. Metrics queries, and also https://grafana.com/docs/tempo/latest/metrics-generator/ I think would do the same thing (I believe OTEL's span metrics connector was inspired by Grafana's one)